test/model/message_test.dart

```diff
@@ -524,24 +531,137 @@ void main() {
-      check(store.messages).deepEquals({1: message});
-      final newMessage = eg.streamMessage(id: 1, content: '<p>bar</p>');
-      final messages = [newMessage];
-      store.reconcileMessages(messages);
-      check(messages).single.identicalTo(message);
-      check(store.messages).deepEquals({1: message});
+    group('fetched message with ID already in store.messages', () {
+      /// Makes a copy of the single message in [MessageStore.messages]
+      /// by round-tripping through [Message.fromJson] and [Message.toJson].
+      ///
+      /// If that message's [StreamMessage.conversation.displayRecipient]
+      /// is null, callers must provide a non-null [displayRecipient]
+      /// to allow [StreamConversation.fromJson] to complete without throwing.
+      Message copyStoredMessage({String? displayRecipient}) {
+        final message = store.messages.values.single;
+
+        final json = message.toJson();
+        if (
+          message is StreamMessage
+          && message.conversation.displayRecipient == null
+        ) {
+          if (displayRecipient == null) throw ArgumentError();
+          json['display_recipient'] = displayRecipient;
+        }
+
+        return Message.fromJson(json);
+      }
+
+      /// Checks if the single message in [MessageStore.messages]
+      /// is identical to [message].
+      void checkStoredMessageIdenticalTo(Message message) {
+        check(store.messages)
+          .deepEquals({message.id: conditionIdentical(message)});
+      }
+
+      test('DM', () async {
+        await prepare();
+        final message = eg.dmMessage(id: 1, from: eg.otherUser, to: [eg.selfUser]);
+
+        store.reconcileMessages([message]);
+        checkStoredMessageIdenticalTo(message);
+        store.reconcileMessages([copyStoredMessage()]);
+        // Not clobbering, because the first call didn't mark stale.
+        checkStoredMessageIdenticalTo(message);
+      });
+
+      group('channel message; chooses correctly whether to clobber the stored version', () {
+        // Exercise the ways we move the message in and out of the "maybe stale"
+        // state. These include reconcileMessage itself, so sometimes we test
+        // repeated calls to that with nothing else happening in between.
+
+        Message checkClobber({Message? withMessageCopy}) {
+          final messageCopy = withMessageCopy ?? copyStoredMessage();
+          store.reconcileMessages([messageCopy]);
+          checkStoredMessageIdenticalTo(messageCopy);
+          return messageCopy;
+        }
+
+        void checkNoClobber({required Message messageBefore}) {
+          store.reconcileMessages([copyStoredMessage()]);
+          checkStoredMessageIdenticalTo(messageBefore);
+        }
+
+        test('various conditions', () async {
+          final channel = eg.stream();
+          await prepare(stream: channel, isChannelSubscribed: true);
+          final message = eg.streamMessage(id: 1, stream: channel);
```

Comment:
```suggestion
          final message = eg.streamMessage(stream: channel);
```

The message ID isn't relevant, right? (Similarly below.)